File: barking_workbook/0x07/BOJ10866.cpp
#include <bits/stdc++.h>
using namespace std;

const int MX = 1000005;
int d[MX*2 + 1];
int head = MX, tail = MX;

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    int N;
    cin >> N;

    while(N--) {
        string s;
        cin >> s;

        if(s == "push_back") {
            int i;
            cin >> i;
            d[tail++] = i;
        } else if(s=="push_front") {
            int i;
            cin >> i;
            d[--head] = i;
        } else if(s=="pop_front") {
            if(head == tail) {
                cout << -1 << '\n';
            } else {
                cout << d[head] << '\n';
                head++;
            }
        } else if(s=="pop_back") {
            if(head == tail) {
                cout << -1 << '\n';
            } else {
                cout << d[tail-1] << '\n';
                tail--;
            }
        } else if(s=="size") {
            cout << tail - head << '\n';
        } else if(s=="empty") {
            if(head == tail) cout << 1 << '\n';
            else cout << 0 << '\n';
        } else if(s=="front") {
            if(head == tail)  cout << -1 << '\n';
            else cout << d[head] << '\n';
        } else {    // back
            if(head == tail)  cout << -1 << '\n';
            else cout << d[tail-1] << '\n';
        }
    }
}

File: coding_test_study/app/src/main/java/coding_test_study/BOJ10866.java
package coding_test_study;

import java.io.*;
import java.util.*;

public class BOJ10866 {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        Deque<Integer> deque = new ArrayDeque<>();

        int n = Integer.parseInt(br.readLine());

        while(n-- > 0) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            String command = st.nextToken();
            switch (command) {
                case "push_front":
                    int x = Integer.parseInt(st.nextToken());
                    deque.addFirst(x);
                    break;
                case "push_back":
                    int y = Integer.parseInt(st.nextToken());
                    deque.addLast(y);
                    break;
                case "pop_front":
                    if (deque.isEmpty()) {
                        System.out.println("-1");
                    } else {
                        System.out.println(deque.pollFirst());
                    }
                    break;
                case "pop_back":
                    if (deque.isEmpty()) {
                        System.out.println("-1");
                    } else {
                        System.out.println(deque.pollLast());
                    }
                    break;
                case "size":
                    System.out.println(deque.size());
                    break;
                case "empty":
                    if (deque.isEmpty()) {
                        System.out.println("1");
                    } else {
                        System.out.println("0");
                    }
                    break;
                case "front":
                    if (deque.isEmpty()) {
                        System.out.println("-1");
                    } else {
                        System.out.println(deque.peekFirst());
                    }
                    break;
                case "back":
                    if (deque.isEmpty()) {
                        System.out.println("-1");
                    } else {
                        System.out.println(deque.peekLast());
                    }
                    break;
            }
        }
        br.close();
    }
}
